cpp/ParseFasm.cpp

```diff
@@ -0,0 +1,414 @@
+#include "antlr4-runtime.h"
+#include "FasmLexer.h"
+#include "FasmParser.h"
+#include "FasmParserVisitor.h"
+
+// External C Interface
+//
+// These functions serialize the FASM parse tree to an easy to parse
+// tag/length/value binary format, where the tag is one byte and
+// the length is 4 bytes, in native endianness (typically little.)
+extern "C" {
+  const char *from_string(const char *in, bool hex);
+  const char *from_file(const char *path, bool hex);
+}
+
+using namespace antlr4;
+using namespace antlrcpp;
+
+// Hex mode is useful for debugging.
+// In this mode, binary values are printed as hex values surrounded by < >
+bool hex_mode = false;
+
+// The Num class provides an << operator that either dumps the raw value
+// or prints the hex value based on the mode. It also can print a tag.
+class Num {
+public:
+
+  // Constructors, with optional tag character
+  Num(char tag, uint32_t num) : num(num), tag(tag) {}
+  Num(uint32_t num) : num(num), tag(0) {}
+
+  uint32_t num;  // The value
+  char tag;  // Tag character
+
+  // The bit width
+  static constexpr int width = sizeof(num) * 8;
+
+};
+
+// Output stream operator for Num.
+// In hex mode, Nums are printed as <XX>, otherwise they are written raw.
```

Comment:
No idea what "written raw" means.

Reply:
I've updated the comment. It means it's copied directly from memory using [`std::ostream::write`](https://en.cppreference.com/w/cpp/io/basic_ostream/write).